Managers/EnemyManager.gd

```diff
@@ -2,7 +2,131 @@ extends Node
+## The reset is handled via signals to be sure to not forget it when changing the battler
+var current_enemy_group_packed: PackedScene = null
+
+## The last recorded lower bound of the current sub-section
+var current_lower_bound: int = enemy_group_distribution.keys()[0][1]
+## The last recorded higher bound of the current sub-section
+var current_higher_bound: int = enemy_group_distribution.keys()[0][2]
+## The last used array of enemy group
+var enemy_group_array: Array[PackedScene]
+## The shuffled version of the `enemy_group_array`
+var shuffled_enemy_group_array: Array[PackedScene] = []
+
+func _ready() -> void:
+	if enemy_group_array.is_empty():
+		enemy_group_array.assign(enemy_group_distribution.values()[0])
+	PhaseManager.on_event_win.connect(_clear_data)
+	PhaseManager.on_defeat.connect(_clear_data)
+
+
+## This calls the inner function that actually chooses the enemy group [br]
+## This is done to save the packed scene in this EnemyManger before it is returned
+func choose_enemy_group() -> PackedScene:
+	if current_enemy_group_packed != null:
+		# if we load from a save
+		return current_enemy_group_packed
+	var packed_enemy_group: PackedScene = _choose_enemy_group_inner()
+	current_enemy_group_packed = packed_enemy_group
+	return packed_enemy_group
+
+## Chooses an enemy group in the following way: [br]
+## - Check if the player is in the same sub-section as the previous time this function was called [br]
+## - Choose an array of enemy group depending on player position if either sub-section changed, or the array to choose from is empty [br]
+## - Shuffle it, pop the last element of that array
+func _choose_enemy_group_inner() -> PackedScene:
+	if current_enemy_group != null:
+		return current_enemy_group_packed
+	if DebugVar.DEBUG_USE_TEST_ENEMY_GROUP:
+		return load("res://Entity/Enemy/EnemyGroup/test_group.tscn") # return a test group
+	var height_percent_position: float = MapManager.get_map_percent_with_player_position()
+	if (current_lower_bound <= height_percent_position && current_higher_bound > height_percent_position):
+		# meaning we stayed in the same sub-section of the map, keep the same array of enemies
+		if shuffled_enemy_group_array.is_empty():
+			shuffled_enemy_group_array = enemy_group_array.duplicate()
+			shuffled_enemy_group_array.shuffle()
+
+		return shuffled_enemy_group_array.pop_back()
+	else:
+		# we entered a new sub-section, search the corresponding array of enemy group
+		for key: Array in enemy_group_distribution:
+			# include lower bound, exclude higher bound
+			if (key[1] <= height_percent_position && key[2] > height_percent_position):
+				# update lower and higher bound
+				current_lower_bound = key[1]
+				current_higher_bound = key[2]
+				var new_enemy_group_array: Array = enemy_group_distribution[key]
+				if new_enemy_group_array.is_empty():
+					break # go to load the test group
+				enemy_group_array.assign(new_enemy_group_array) # assign performs a copy
```

Comment:
We don't need to deep copy the array, since we are not modifying the elements of the array (only instantiating them); we are only modifying the arrays themselves.
Indeed, I thought assign would perform a copy, but seems like it does not. It doesn't really matter if we do a copy of `enemy_group_array` either, since we do not modify this array; I can leave a comment for the future in case we need it